public/js/services/product-service.js

```diff
@@ -1,6 +1,38 @@
 class ProductService {
-  findProducts() {
-    return $.getJSON('/api/products.json')
+  async findProducts(category = '', search = '') {
+    const products = await $.getJSON('/api/products.json')
+    let filteredProducts = []
+
+    if (category !== '' && search !== '') {
+      //If category and search filters are present.
+      products.forEach((product) => {
+        if (
+          category === product.category &&
+          product.name.toLowerCase().includes(search.toLowerCase()) === true
+        ) {
+          filteredProducts.push(product)
+        }
+      })
+      return filteredProducts
+    }
+
+    if (category !== '') {
+      // If category is present, but not search filter.
+      products.forEach((product) => {
+        if (category === product.category) filteredProducts.push(product)
+      })
+      return filteredProducts
+    }
+
+    if (search !== '') {
+      // If search filter is present, but not category.
+      products.forEach((product) => {
+        if (product.name.toLowerCase().includes(search.toLowerCase()))
+          filteredProducts.push(product)
+      })
+      return filteredProducts
+    }
+    return products
```

Comment:
Let's see if we can generalize this code.

We can define these at the top

```js
const CategoryFilter = function(category) {
  return (product) => product.category === category
}

const SearchFilter = function(search) {
  return (product) => product.name.toLowerCase().includes(search.toLowerCase())
}
```

And simplify the filtering to something like
```js
async findProducts(category = '', search = '') {
  const products = await $.getJSON('/api/products.json')
  const filters = []
  
  if (category) {
    filters.push(CategoryFilter(category))
  }
  
  if (search) {
    filters.push(SearchFilter(search))
  }
  
  for (let filter of filters) {
    products = products.filter(filter)
  }

  return products
}
```

That way if we were to add more kinds of filters in the future it's trivial to set it up